### PR TITLE
Add feature to Browse commons via app

### LINKED ISSUE #1221: Allow users to browse Commons from app
**Summary:** 

 My suggestion is to enable the user to access the Mobile version of Wikimedia Commons  Website  within their app. This would give an integrated feel to the user. 
I noticed that this has been the main concern with people in playstore reviews.
Sorry if this issue has already been discussed

**Steps to reproduce:** 
Adding a new activity which would display the mobile version of Wikimedia Commons app using a webView.

### PR DESCRIPTION
**Pull request by ujjwalagrawal17**
## Title (required)
- Fixes #1221 (Allows users to browse commons from app)  

## Description (required)

- Improvements in category images activity (Changed it to explore activity)
- added search activity with 2 fragments (one for searching images, other for categories)
- Added recent Search history and delete recent search hostory feature
- Added a category details page (includes media, subcategories, parent categories inside that category)

## Tests performed (required)

Manually Tested on API 25 & MOTO G5S+, with {ProdDebug variant}.

## Screenshots showing what changed (optional)
<table>
<tr>
<td>

screenshot_20180714-172916

</td>
<td>

screenshot_20180714-172913

</td>
<td>

screenshot_20180714-172907

</td>
</tr>
<tr>
<td>

screenshot_20180714-172901

</td>
<td>

screenshot_20180714-172856

</td>
<td>

screenshot_20180714-172839

</td>
</tr>
<tr>
<td>

screenshot_20180714-172829

</td>
<td>

screenshot_20180714-172820

</td>
</tr>
</table>

### CODE REVIEW

**Comment by ujjwalagrawal17** [comment]
@misaochan @maskaravivek @neslihanturan @nicolas-raoul Can you review the PR.

File: app/src/main/java/fr/free/nrw/commons/di/CommonsApplicationModule.java [hunk 1]
```diff
@@ -53,6 +54,13 @@ public ContentProviderClient provideCategoryContentProviderClient(Context contex
         return context.getContentResolver().acquireContentProviderClient(CATEGORY_AUTHORITY);
     }
 
+    @Provides
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/di/CommonsApplicationModule.java]: Javadoc

File: app/src/main/java/fr/free/nrw/commons/explore/categories/SearchCategoriesAdapterFactory.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package fr.free.nrw.commons.explore.categories;
+
+import com.pedrogomez.renderers.ListAdapteeCollection;
+import com.pedrogomez.renderers.RVRendererAdapter;
+import com.pedrogomez.renderers.RendererBuilder;
+
+import java.util.Collections;
+import java.util.List;
+
+public class SearchCategoriesAdapterFactory {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/explore/categories/SearchCategoriesAdapterFactory.java]: Javadoc

File: app/src/main/java/fr/free/nrw/commons/explore/images/SearchImagesAdapterFactory.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import com.pedrogomez.renderers.RVRendererAdapter;
+import com.pedrogomez.renderers.RendererBuilder;
+
+import java.util.Collections;
+import java.util.List;
+
+import fr.free.nrw.commons.Media;
+
+
+class SearchImagesAdapterFactory {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/explore/images/SearchImagesAdapterFactory.java]: Javadoc

File: app/src/main/java/fr/free/nrw/commons/explore/recentsearches/RecentSearchesContentProvider.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+import fr.free.nrw.commons.data.DBOpenHelper;
+import fr.free.nrw.commons.di.CommonsDaggerContentProvider;
+import timber.log.Timber;
+
+import static android.content.UriMatcher.NO_MATCH;
+import static fr.free.nrw.commons.explore.recentsearches.RecentSearchesDao.Table.ALL_FIELDS;
+import static fr.free.nrw.commons.explore.recentsearches.RecentSearchesDao.Table.COLUMN_ID;
+import static fr.free.nrw.commons.explore.recentsearches.RecentSearchesDao.Table.TABLE_NAME;
+
+public class RecentSearchesContentProvider extends CommonsDaggerContentProvider {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/explore/recentsearches/RecentSearchesContentProvider.java]: Javadoc

File: app/src/main/java/fr/free/nrw/commons/explore/recentsearches/RecentSearchesDao.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+
+
+public class RecentSearchesDao {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/explore/recentsearches/RecentSearchesDao.java]: Javadoc

File: app/src/main/java/fr/free/nrw/commons/explore/recentsearches/RecentSearchesDao.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        } finally {
+            if (cursor != null) {
+                cursor.close();
+            }
+            db.release();
+        }
+        return items;
+    }
+
+    @NonNull
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/explore/recentsearches/RecentSearchesDao.java]: javadoc

File: app/src/main/java/fr/free/nrw/commons/theme/NavigationBaseActivity.java [hunk 7]
```diff
@@ -82,6 +82,11 @@ public void initBackButton() {
         toggle.setToolbarNavigationClickListener(v -> onBackPressed());
     }
 
+    public void forceInitBackButton() {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/theme/NavigationBaseActivity.java]: Javadoc

**Comment by ujjwalagrawal17** [comment]
@nicolas-raoul Ok I will add the Javadocs for these in a PR to browse-branch.

File: app/src/main/java/fr/free/nrw/commons/category/CategoryDetailsActivity.java [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+     * @param categoryName Name of the category for displaying its details
+     */
+    public static void startYourself(Context context, String categoryName) {
+        Intent intent = new Intent(context, CategoryDetailsActivity.class);
+        intent.addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
+        intent.putExtra("categoryName", categoryName);
+        context.startActivity(intent);
+    }
+
+    @Override
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryDetailsActivity.java]: There are still many methods without Javadoc it seems, such as these ones.
Please add Javadoc to all of them, thanks! :-)

Comment (ujjwalagrawal17) [reply app/src/main/java/fr/free/nrw/commons/category/CategoryDetailsActivity.java]: @nicolas-raoul Ok, I will search and do it for all.

Comment (ujjwalagrawal17) [reply app/src/main/java/fr/free/nrw/commons/category/CategoryDetailsActivity.java]: @nicolas-raoul I have added it for most of the methods now. Can you tell if we need it in any other method/class?

**Comment by ujjwalagrawal17** [comment]
@nicolas-raoul Can we merge #1737 it will fix the conflicts between 2 branches.

**Comment by ujjwalagrawal17** [comment]
@nicolas-raoul @neslihanturan Also, can you both review it once for any other changes.

File: app/src/main/java/fr/free/nrw/commons/category/CategoryImagesActivity.java [hunk 9]
```diff
@@ -102,17 +115,32 @@ public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
             FragmentManager supportFragmentManager = getSupportFragmentManager();
             supportFragmentManager
                     .beginTransaction()
-                    .replace(R.id.fragmentContainer, mediaDetails)
+                    .hide(supportFragmentManager.getFragments().get(supportFragmentManager.getBackStackEntryCount()))
+                    .add(R.id.fragmentContainer, mediaDetails)
```
Comment (neslihanturan) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryImagesActivity.java]: Aparently there is a reason you used "add" isdeat of "replace" here. Can you plase note this reason for further developers who will work on this code?

File: app/src/main/java/fr/free/nrw/commons/category/CategoryImagesListFragment.java [hunk 10]
```diff
@@ -48,9 +50,9 @@
     TextView statusTextView;
     @BindView(R.id.loadingImagesProgressBar) ProgressBar progressBar;
     @BindView(R.id.categoryImagesList) GridView gridView;
-
+    @BindView(R.id.parentLayout) RelativeLayout parentLayout;
     private boolean hasMoreImages = true;
-    private boolean isLoading;
+    private boolean isLoading=true;
```
Comment (neslihanturan) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryImagesListFragment.java]: Spaces before and after "=" please

File: app/src/main/java/fr/free/nrw/commons/category/CategoryImagesListFragment.java [hunk 11]
```diff
@@ -213,6 +223,10 @@ private void handleSuccess(List<Media> collection) {
         if(gridAdapter == null) {
             setAdapter(collection);
         } else {
+            if (gridAdapter.containsAll(collection)){
+                hasMoreImages = false;
```
Comment (neslihanturan) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryImagesListFragment.java]: Space before { please

**Comment by neslihanturan** [comment]
Looks very clean @ujjwalagrawal17 , nice job!

**Comment by ujjwalagrawal17** [comment]
@neslihanturan @nicolas-raoul can we merge it now ??

**Comment by misaochan** [comment]
Go for it! Thanks everyone who participated in this. :)

**Comment by neslihanturan** [comment]
Congratz!

**Comment by domdomegg** [comment]
Can this branch be deleted now?

**Comment by misaochan** [comment]
I think so. Can @ujjwalagrawal17 confirm?

**Pull request merged**